navigation/image_database.h

```diff
@@ -0,0 +1,173 @@
+#include <vector>
+
+namespace BoBRobotics {
+namespace Navigation {
+using namespace units::length;
+using namespace units::angle;
+
+class ImageDatabase
+{
+public:
+    struct Entry {
+        Vector3<millimeter_t> position;
+        degree_t heading;
+        std::string path;
+
+        cv::Mat load() const
+        {
+            return cv::imread(path);
+        }
+
+        cv::Mat loadGreyscale() const
+        {
+            return cv::imread(path, cv::IMREAD_GRAYSCALE);;
+        }
+    };
+
+    template<class T>
+    ImageDatabase(T &&databaseName)
+      : m_Path(std::forward<T>(databaseName))
+    {
+        const auto metadataPath = m_Path / "metadata.csv";
+
+        // If we don't have metadata, it's an empty database
+        if (!metadataPath.exists()) {
+            // Make sure we have a directory to save into
+            filesystem::create_directory(m_Path);
+            return;
+        }
+
+        // Otherwise parse metadata file
+        std::ifstream file(metadataPath.str());
+        std::string line;
+        std::vector<std::string> fields;
+        std::getline(file, line); // skip first line
+        while (!file.eof()) {
+            // Values are separated by commas
+            fields.clear();
+            std::getline(file, line);
+            size_t start = 0;
+            for (size_t end = 0; end < line.size(); ) {
+                if (line[end] == ',') {
+                    fields.push_back(line.substr(start, end - start));
+                    start = end + 1;
+
+                    // Skip whitespace
+                    while (start < line.size() && line[start] == ' ') {
+                        start++;
+                    }
+                    end = start;
+                } else {
+                    end++;
+                }
+            }
+            fields.push_back(line.substr(start));
+            if (fields.size() < 5) {
+                break;
+            }
+
+            // Save details to vector
+            Entry entry {
+                {
+                    millimeter_t(std::stod(fields[0])),
+                    millimeter_t(std::stod(fields[1])),
+                    millimeter_t(std::stod(fields[2]))
+                },
+                degree_t(std::stod(fields[3])),
+                (m_Path / fields[4]).str()};
+            m_Database.push_back(entry);
+        }
+    }
+
+    ~ImageDatabase()
+    {
+        // If we need to write metadata and haven't yet done so, do it now
+        if (m_DirtyFlag) {
+            saveMetadata();
+        }
+    }
+
+    const filesystem::path &getPath() const { return m_Path; }
+
+    const Entry &operator[](size_t i) const { return m_Database[i]; }
+    auto begin() const { return m_Database.cbegin(); }
+    auto end() const { return m_Database.cend(); }
+    size_t size() const { return m_Database.size(); }
+
+    void saveImage(cv::Mat &frame, millimeter_t x, millimeter_t y, millimeter_t z, degree_t heading, bool isRoute)
```

Comment:
Maybe this would be better named ``addImage``?

Reply:
Good shout.